### DISCUSSION TITLE
Reverse proxy to magento store

### DISCUSSION THREAD

**Discussion by f1-outsourcing** [OpenMage/magento-lts#3426] [Questions & Answers]
I have a magento store on shop.example.com and I would like to publish it on www.test.com/shop. With apache proxypass I able to get the home page. But magento seems to put full urls everywhere

```
<script type="text/javascript" src="https://shop.example.com/js/prototype/prototype.js"></script>
<script type="text/javascript" src="https://shop.example.com/js/lib/jquery/jquery-1.12.1.min.js"></script>
<script type="text/javascript" src="https://shop.example.com/js/lib/jquery/noconflict.js"></script>
<script type="text/javascript" src="https://shop.example.com/js/lib/ccard.js"></script>
<script type="text/javascript" src="https://shop.example.com/js/prototype/validation.js"></script>
```

Is it possible to put Magento in a relative url mode?

**Comment by f1-outsourcing** [comment]
This seems to get me quite far, except with logging in, then it goes back to original domain


```
     23 SSLProxyEngine On
     24 SSLProxyVerify none
     25 SSLProxyCheckPeerCN off
     26 SSLProxyCheckPeerName off
     27 SSLProxyCheckPeerExpire off
     28 <Location /shop/>
     29         SetOutputFilter  proxy-html
     30         ProxyPass        "https://shop.example.com/"
     31         ProxyPassReverse "https://shop.example.com/"
     32
     33         ProxyHTMLURLMap https://shop.example.com/ /shop/
     34
     35         Options +Indexes +ExecCGI +FollowSymLinks -MultiViews
     36         Order Allow,Deny
     37         Allow from all
     38 </Location>

```

**Reply by elidrissidev** [reply]
The issue with login is because the cookies are still set to the original domain, so the domain in `Set-Cookie` headers need to be modified on-the-fly.

**Comment by f1-outsourcing** [comment]
Yes I had indeed issues with the cookies and my testing frontend did not have ssl enabled, so the cookies were not even set in the browser, ended up using this. 

```
     35
     36 SSLProxyEngine On
     37 SSLProxyVerify none
     38 SSLProxyCheckPeerCN off
     39 SSLProxyCheckPeerName off
     40 SSLProxyCheckPeerExpire off
     41 <Location /shop>
     42         Define proxyhost "shop.example.com"
     43         #Define proxyhost "store.example.com"
     44
     45         SetOutputFilter  proxy-html
     46         ProxyHTMLExtended On
     47
     48         ProxyPass        "https://${proxyhost}"
     49         ProxyPassReverse "https://${proxyhost}"
     50
     51         ProxyPassReverseCookieDomain    "${proxyhost}" "test.com"
     52         ProxyPassReverseCookiePath      "/" "/shop"
     53
     54         Header edit Set-Cookie ^(.*)$ $1;HttpOnly;Secure;SameSite=None
     55
     56         ProxyHTMLURLMap https://${proxyhost} /shop
     57
     58         Options +ExecCGI +FollowSymLinks -MultiViews
     59         Order Allow,Deny
     60         Allow from all
     61 </Location>
     62
```

I thought the ProxyPassReverseCookieDomain   would be enough but somehow I still need Header edit Set-Cookie

**Comment by f1-outsourcing** [comment]
I have more or less working config with a reverse proxy. I can log in, I have a cart. The one page checkout lists my address. Only the last button "place order" keeps loading.
When I look at the source of this page, I have still the internal host domain configured which was not changed.

```
<html lang="en" id="top" class="no-js"><head><
...
...
...
/css?family=Raleway:300,400,500,700,600&amp;display=swap"><link rel="stylesheet" type="text/css" href="/skin/frontend/rwd/adfasdf/css/styles.css" media="all"><link rel="stylesheet" type="text/css" href="/skin/frontend/rwd/default/css/madisonisland.css" media="all"><script type="text/javascript">
//<![CDATA[
Mage.Cookies.path     = '/';
Mage.Cookies.domain   = '.shop.example.com';
//]]>
```

Maybe it would be good in newer releases to code this in such a way it would be compatible with reverse proxies that change host names.



#3426

**Reply by f1-outsourcing** [reply]
```
Mage.Cookies.domain   = '.shop.example.com';
```

It is not related to this, this is on multiple pages and seems to be corrected by apache's cookie manipulating settings.

**Comment by f1-outsourcing** [comment]
If I print the url from js/prototype/prototype.js to the console, and continue through the one page checkout, I am getting these results. First the correct urls (external) are being used.

```
https://shop.goodurl.com/en_gb/checkout/onepage/saveBilling/

https://shop.goodurl.com/en_gb/checkout/onepage/progress/

https://shop.goodurl.com/en_gb/checkout/onepage/savePayment/

https://shop.goodurl.com/en_gb/checkout/onepage/progress/

```

But the payment is going to the wrong  unchanged  original  url
```
https://shop.badurl.com/en_gb/checkout/onepage/saveOrder/
```

**Reply by f1-outsourcing** [reply]
getting this cors error

```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://xxxx/en_gb/checkout/onepage/saveOrder/. (Reason: CORS request external redirect not allowed).
```

**Comment by kiatng** [comment]
> I have a magento store on shop.example.com and I would like to publish it on [www.test.com/shop](http://www.test.com/shop).

Why not create a new store and set the base URLs to http://www.test.com/shop?

**Reply by f1-outsourcing** [reply]
I think if you create a new store, you are also forced to create new store views, and those codes are (still?) unique throughout the install (and not unique in combination with eg website,store). So I am having an 'en_gb' as storeview name in the url, and what can I than use else en_gb2 in the url? en_gb3 in url? Just does not look ok.
If you use the proxy, you can also force people in a store view which they can't exit. I think doing this with the proxy is a nicer solution. It sounds more logical to fix it like this.
( it is has been a while since I looked at this, so I am a bit limited with knowledge from the top of my head ;) )

**Reply by kiatng** [reply]
The configuration is minimum, you can test it easily.
![image](https://github.com/OpenMage/magento-lts/assets/1106470/6121fcc5-1d24-4615-b812-e458d2a09c8c)

The URL won't show store view.

There will be a store switcher rendered in the header, which you can removed.
![image](https://github.com/OpenMage/magento-lts/assets/1106470/325dbca5-8949-48c6-a321-fe9131091330)

Just delete the child block `'store_language'`.
![image](https://github.com/OpenMage/magento-lts/assets/1106470/fc204098-6546-4c2b-9e72-3c0e24bf49cd)

**Comment by f1-outsourcing** [comment]
this url is url is generating a json response which is not being processed by ProxyHTMLURLMap

/en_gb/checkout/onepage/savePayment/

This can be fixed by using the apache filter module
```
    105 <LocationMatch /.._../checkout/onepage/savePayment>
    106   FilterDeclare MYFILTER
    107   FilterProvider MYFILTER SUBSTITUTE "%{Content_Type} = 'application/json'"
    108   FilterChain MYFILTER
    109   Substitute "s!${proxyhost}!${defaulthost}!i"
    110 </LocationMatch>
```
